language/bytecode-verifier/src/borrow_graph.rs

```diff
@@ -0,0 +1,230 @@
+        checked_precondition!(self.0.contains_key(&to));
+        checked_precondition!(self.0[&to].is_empty());
+        checked_precondition!(label.len() <= 1);
+
+        let new_edge = Edge {
+            edge_type: EdgeType::Strong,
+            label: label.clone(),
+            to,
+        };
+        self.0.remove(&to).unwrap();
+        if label.is_empty() {
+            let from_edge_set = self.0.remove(&from).unwrap();
+            let mut new_from_edge_set = BTreeSet::new();
+            new_from_edge_set.insert(new_edge);
+            self.0.insert(from, new_from_edge_set);
+            self.0.insert(to, from_edge_set);
+        } else {
+            let mut edge_set: Vec<Edge<T>> = self.0[&from]
+                .iter()
+                .filter(|x| !x.label.is_empty() && x.label[0] == label[0])
+                .cloned()
+                .collect();
+            let from_edge_set = self.0.get_mut(&from).unwrap();
+            edge_set.iter().for_each(|x| {
+                from_edge_set.remove(x);
+            });
+            from_edge_set.insert(new_edge);
+            edge_set.iter_mut().for_each(|x| {
+                x.label.remove(0);
+            });
+            self.0.insert(to, edge_set.into_iter().collect());
+        }
+    }
+
+    /// removes `nonce` and appropriately concatenates each incoming edge with each outgoing edge of `nonce`
+    pub fn remove_nonce(&mut self, nonce: Nonce) {
+        let nonce_edge_set = self.0.remove(&nonce).unwrap();
+        let mut removed_edges = BTreeMap::new();
+        for (n, es) in &self.0 {
+            removed_edges.insert(
+                n.clone(),
+                es.iter()
+                    .filter(|x| x.to == nonce)
+                    .cloned()
+                    .collect::<Vec<_>>(),
+            );
+        }
+        for (n, es) in &removed_edges {
+            es.iter().for_each(|removed_edge| {
+                let n_edge_set_ref = self.0.get_mut(n).unwrap();
+                n_edge_set_ref.remove(removed_edge);
+                nonce_edge_set.iter().for_each(|nonce_edge| {
+                    if removed_edge.edge_type == EdgeType::Strong {
+                        let mut new_label = vec![];
+                        new_label.append(&mut removed_edge.label.clone());
+                        new_label.append(&mut nonce_edge.label.clone());
+                        let edge = Edge {
+                            edge_type: nonce_edge.edge_type,
+                            label: new_label,
+                            to: nonce_edge.to,
+                        };
+                        n_edge_set_ref.insert(edge);
+                    } else {
+                        let edge = Edge {
+                            edge_type: EdgeType::Weak,
+                            label: removed_edge.label.clone(),
+                            to: nonce_edge.to,
+                        };
+                        n_edge_set_ref.insert(edge);
+                    }
```

Comment:
This should be factored out into a separate function for splicing out the removed id. I think it will make this code a lot simpler

Reply:
We will revisit when we move towards a unified borrow graph implementation.